macq/extract/locm.py

```diff
@@ -710,114 +754,122 @@ def _step7(
-                        end_fluent_temp.param_sorts,
-                        [ap.pos - 1] + bound_param_inds,
+                start_fluent_name = f"sort{sort}_state{start_state}"
+                if start_fluent_name not in fluents[ap.action.name]:
+                    start_fluent = LearnedLiftedFluent(
+                        start_fluent_name,
+                        param_sorts=[sort_str],
+                        param_act_inds=[ap.pos - 1],
                     )
-                    fluents[sort][end_state] = end_fluent
-                    actions[ap.action.name].update_add(end_fluent)
+                    fluents[ap.action.name][start_fluent_name] = start_fluent
 
-        fluents = set(fluent for sort in fluents.values() for fluent in sort.values())
-        actions = set(actions.values())
+                start_fluent = fluents[ap.action.name][start_fluent_name]
+
+                if (
+                    sort in state_params_to_hyps
+                    and start_state in state_params_to_hyps[sort]
+                ):
+                    for param in state_params_to_hyps[sort][start_state]:
+                        psort = None
+                        pind = None
+                        for hyp in state_params_to_hyps[sort][start_state][param]:
+                            if hyp.C == ap:
+                                assert psort is None or psort == hyp.G_
+                                assert pind is None or pind == hyp.l_
+                                psort = hyp.G_
+                                pind = hyp.l_
+                        assert psort is not None
+                        assert pind is not None
+                        start_fluent.param_sorts.append(f"sort{psort}")
+                        start_fluent.param_act_inds.append(pind - 1)
+
+                a.update_precond(start_fluent)
+
+                if end_state != start_state:
+                    end_fluent_name = f"sort{sort}_state{end_state}"
+                    if end_fluent_name not in fluents[ap.action.name]:
+                        end_fluent = LearnedLiftedFluent(
+                            end_fluent_name,
+                            param_sorts=[sort_str],
+                            param_act_inds=[ap.pos - 1],
+                        )
+                        fluents[ap.action.name][end_fluent_name] = end_fluent
+
+                    end_fluent = fluents[ap.action.name][end_fluent_name]
+
+                    if (
+                        sort in state_params_to_hyps
+                        and end_state in state_params_to_hyps[sort]
+                    ):
+                        for param in state_params_to_hyps[sort][end_state]:
+                            psort = None
+                            pind = None
+                            for hyp in state_params_to_hyps[sort][end_state][param]:
+                                if hyp.B == ap:
+                                    assert psort is None or psort == hyp.G_
```

Comment:
do we still want all these assertions in the final implementation? shouldn't matter if they pass I suppose

Reply:
Ya, my preference would be to leave them -- useful for detecting regressions.